File: python_backend/app/models/building_genome.py
from __future__ import annotations

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class Archetype(BaseModel):
    id: str = "generic"
    confidence: float = 0.0


class Topology(BaseModel):
    layout: str = "rectangular"     # rectangular/circular/linear/radial/freeform
    composition: str = "single"     # single/cluster/chain/grid
    axis: str = "none"              # centered/axial/none
    levels: str = "mixed"           # horizontal/vertical/mixed


class Structure(BaseModel):
    type: str = "hybrid"            # solid/frame/hybrid/suspended
    massiveness: Optional[float] = None
    voidRatio: Optional[float] = None
    supports: Optional[str] = None  # central/distributed


class Form(BaseModel):
    repetition: Optional[str] = None
    progression: Optional[str] = None
    curvature: Optional[str] = None
    rhythm: Optional[str] = None


class Symmetry(BaseModel):
    type: str = "none"              # none/bilateral/radial/grid
    order: Optional[int] = None
    mirror: Optional[bool] = None


class Materials(BaseModel):
    primary: Optional[str] = None   # stone/wood/earth/metal/glass/mixed
    secondary: Optional[str] = None
    accent: Optional[str] = None
    textureBias: Optional[str] = None
    extra: Optional[Dict[str, Any]] = None


class CulturalStyle(BaseModel):
    region: Optional[str] = None
    era: Optional[str] = None
    keywords: Optional[List[str]] = None


class Constraints(BaseModel):
    maxHeight: Optional[int] = None
    respectTerrain: Optional[bool] = None
    insideSelectionOnly: Optional[bool] = None
    noModifyZones: Optional[List[str]] = None


class AIHints(BaseModel):
    reference: Optional[str] = None
    priority: Optional[List[str]] = None
    avoid: Optional[List[str]] = None
    extra: Optional[Dict[str, Any]] = None


class BuildingGenome(BaseModel):
    """
    BuildingGenome v1：FormaCraft 建筑 DNA 标准
    - 不描述具体方块/patch
    - 建议挂载方式：BuildingSpec.extra["genome"] = BuildingGenome(dict)
    """
    genomeVersion: str = Field(default="1.0")

    archetype: Archetype = Field(default_factory=Archetype)
    topology: Topology = Field(default_factory=Topology)
    structure: Structure = Field(default_factory=Structure)
    form: Form = Field(default_factory=Form)
    symmetry: Symmetry = Field(default_factory=Symmetry)

    modules: List[str] = Field(default_factory=list)
    materials: Materials = Field(default_factory=Materials)
    culturalStyle: CulturalStyle = Field(default_factory=CulturalStyle)
    constraints: Constraints = Field(default_factory=Constraints)
    aiHints: AIHints = Field(default_factory=AIHints)

File: src/main/java/com/formacraft/common/genome/BuildingGenome.java
package com.formacraft.common.genome;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * BuildingGenome v1：建筑 DNA 标准（跨 AI/后端/前端/执行层共用）
 *
 * 注意：v1 不描述具体 block/patch，只描述空间逻辑与生成倾向。
 * 建议挂载方式：BuildingSpec.extra.genome = BuildingGenome（JSON object）
 */
public class BuildingGenome {
    public String genomeVersion = "1.0";

    public Archetype archetype = new Archetype();
    public Topology topology = new Topology();
    public Structure structure = new Structure();
    public Form form = new Form();
    public Symmetry symmetry = new Symmetry();

    public List<String> modules = new ArrayList<>();
    public Materials materials = new Materials();
    public CulturalStyle culturalStyle = new CulturalStyle();
    public Constraints constraints = new Constraints();
    public AIHints aiHints = new AIHints();

    /** 原型识别（强形象/地标） */
    public static class Archetype {
        public String id = "generic";
        public double confidence = 0.0;
    }

    /** 空间拓扑 */
    public static class Topology {
        public String layout = "rectangular";     // rectangular/circular/linear/radial/freeform
        public String composition = "single";     // single/cluster/chain/grid
        public String axis = "none";              // centered/axial/none
        public String levels = "mixed";           // horizontal/vertical/mixed
    }

    /** 结构逻辑 */
    public static class Structure {
        public String type = "hybrid";            // solid/frame/hybrid/suspended
        public Double massiveness = null;         // 0~1
        public Double voidRatio = null;           // 0~1
        public String supports = null;            // central/distributed
    }

    /** 形态语法 */
    public static class Form {
        public String repetition = null;          // none/horizontal/vertical/radial
        public String progression = null;         // uniform/tapering/stepping/upward
        public String curvature = null;           // straight/curved/mixed
        public String rhythm = null;              // regular/segmented/irregular
    }

    /** 对称 */
    public static class Symmetry {
        public String type = "none";              // none/bilateral/radial/grid
        public Integer order = null;              // for radial
        public Boolean mirror = null;
    }

    /** 材料语义（非具体方块） */
    public static class Materials {
        public String primary = null;             // stone/wood/earth/metal/glass/mixed
        public String secondary = null;
        public String accent = null;
        public String textureBias = null;         // rough/smooth/polished/aged
        public Map<String, Object> extra = null;  // future-proof
    }

    /** 文化/风格标签 */
    public static class CulturalStyle {
        public String region = null;
        public String era = null;
        public List<String> keywords = null;
    }

    /** 硬约束 */
    public static class Constraints {
        public Integer maxHeight = null;
        public Boolean respectTerrain = null;
        public Boolean insideSelectionOnly = null;
        public List<String> noModifyZones = null;
    }

    /** 仅给 AI 看 */
    public static class AIHints {
        public String reference = null;
        public List<String> priority = null;
        public List<String> avoid = null;
        public Map<String, Object> extra = null;
    }
}


